overlord/snapstate/refresh_control.go

```diff
@@ -0,0 +1,173 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapstate
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/mount"
+	"github.com/snapcore/snapd/overlord/ifacestate/ifacerepo"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+var gateAutoRefreshHookName = "gate-auto-refresh"
+
+func affectedByRefresh(st *state.State, updates []*snap.Info) ([]string, error) {
+	all, err := All(st)
+	if err != nil {
+		return nil, err
+	}
+
+	byBase := make(map[string][]string)
+	for name, snapSt := range all {
+		if !snapSt.Active {
+			delete(all, name)
+			continue
+		}
+		inf, err := snapSt.CurrentInfo()
+		if err != nil {
+			return nil, err
+		}
+		// optimize: do not consider snaps that don't have gate-auto-refresh hook.
+		if inf.Hooks[gateAutoRefreshHookName] == nil {
+			delete(all, name)
+			continue
+		}
+
+		base := inf.Base
+		if inf.Base == "" {
+			base = "core"
+		}
+		byBase[base] = append(byBase[base], inf.InstanceName())
+	}
+
+	affected := make(map[string]bool)
+
+	for _, up := range updates {
+		// add self
+		if all[up.InstanceName()] != nil {
+			affected[up.InstanceName()] = true
+		}
+
+		// snaps that can trigger reboot
+		// XXX: gadget refresh doesn't always require reboot, refine this
+		if up.Type() == snap.TypeKernel || up.Type() == snap.TypeGadget {
+			for _, snapSt := range all {
+				affected[snapSt.InstanceName()] = true
+			}
+			continue
+		}
```

Comment:
on core we also need to consider the boot base (base in model) as its own thing that affects all snaps

Reply:
Done.